File: RaftSwiftGRPC/Sources/RaftCore/Node/RaftNodeApplication.swift
/// Protocol for a Raft node application
public protocol RaftNodeApplication: Sendable {
    /// Initializes the server
    /// - Parameters:
    ///   - id: The ID of this server
    ///   - port: The port to listen on for incoming connections
    ///   - peers: The list of peers
    init(id: Int, port: Int, peers: [Peer])

    /// Starts the node
    func serve() async throws
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/node/RaftNodeApplication.kt
package raft.node

import raft.utils.types.Peer

/**
 * Abstract class for a Raft node application.
 * This class provides the basic functionality for a Raft node application.
 */
abstract class RaftNodeApplication {
    /**
     * The own peer.
     */
    val ownPeer: Peer
    /**
     * The list of peers.
     */
    val peers: MutableList<Peer>
    /**
     * The persistence layer.
     */
    val persistence: RaftNodePersistence

    /**
     * Constructor for a Raft node application.
     *
     * @param ownPeer The own peer.
     * @param peers The list of peers.
     * @param persistence The persistence layer.
     */
    constructor(ownPeer: Peer, peers: MutableList<Peer>, persistence: RaftNodePersistence) {
        this.ownPeer = ownPeer
        this.peers = peers
        this.persistence = persistence
    }

    /**
     * Starts the Raft node application.
     */
    abstract fun serve()
}